### ISSUE TITLE
BUG in __split_pgfrag: unable to handle kernel NULL pointer dereference

### ISSUE THREAD

**Issue by keshonok** [natsys/tempesta#436]
This has happened in one of the runs of the `ab` benchmark. Difficult to reproduce without a special test request that consists of specially crafted SKBs.
```
[52016.303192] BUG: unable to handle kernel NULL pointer dereference at 0000000000000094
[52016.311901] IP: [<ffffffffa06bccd9>] __split_pgfrag+0x5d9/0x930 [tempesta_fw]
[52016.319916] PGD 0
[52016.322761] Oops: 0000 [#1] SMP
[52016.326743] Modules linked in: tfw_sched_rr(O) tfw_sched_http(O) tfw_sched_hash(O) tempesta_fw(O) tempesta_db(O) nfsd auth_rpcgss oid_registry nfs_acl nfs lockd grace fscache sunrpc ast ttm drm_kms_helper intel_powerclamp drm snd_pcm coretemp crct10dif_pclmul crc32_pclmul joydev snd_timer snd aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper i2c_algo_bit psmouse iTCO_wdt iTCO_vendor_support evdev soundcore syscopyarea sysfillrect cryptd lpc_ich sysimgblt pcspkr mfd_core serio_raw i2c_i801 tpm_tis tpm i2c_core button acpi_cpufreq processor i7core_edac thermal_sys edac_core shpchp ipmi_watchdog ipmi_si ipmi_poweroff ipmi_devintf ipmi_msghandler autofs4 ext4 crc16 mbcache jbd2 sg sr_mod cdrom sd_mod hid_generic usbhid hid ehci_pci uhci_hcd ahci libahci ehci_hcd ixgbe e1000e mdio usbcore libata
[52016.404040]  crc32c_intel dca ptp scsi_mod usb_common pps_core [last unloaded: tempesta_db]
[52016.411965] CPU: 1 PID: 13 Comm: ksoftirqd/1 Tainted: G        W  O    4.1.12+ #1
[52016.420184] Hardware name: Quanta                 S89/QR1-890/QSSC-890  /S89/QR1-890/QSSC-890  , BIOS S89_3A10 10/21/2010
[52016.431907] task: ffff8800592aa8d0 ti: ffff8800592b0000 task.ti: ffff8800592b0000
[52016.440145] RIP: 0010:[<ffffffffa06bccd9>]  [<ffffffffa06bccd9>] __split_pgfrag+0x5d9/0x930 [tempesta_fw]
[52016.450528] RSP: 0018:ffff8800592b33e8  EFLAGS: 00010246
[52016.456637] RAX: 0000000000000000 RBX: ffff88004db37600 RCX: 00000000000000c3
[52016.464596] RDX: ffff88004db37800 RSI: 0000000000000000 RDI: ffff88004db37600
[52016.472563] RBP: ffff8800592b3448 R08: ffff88004db37ac0 R09: 0000000000000001
[52016.480542] R10: 0000000000000001 R11: 0000000000000000 R12: 00000000000000c3
[52016.488535] R13: 00000000ffffffff R14: ffff88004db37af0 R15: 0000000000000000
[52016.496533] FS:  0000000000000000(0000) GS:ffff88007a000000(0000) knlGS:0000000000000000
[52016.505469] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[52016.512061] CR2: 0000000000000094 CR3: 0000000001c0c000 CR4: 00000000000006e0
[52016.520049] Stack:
[52016.522918]  00000000000000b0 ffff88004db374c0 00000003592aa8d0 0000000000000000
[52016.531279]  0000000000000000 ffffffff82fecea0 ffffffff828fdb68 00000000000000c3
[52016.539640]  ffff88004db37600 ffffffffffffffff ffff8800592b34b8 ffff88004db37800
[52016.548007] Call Trace:
[52016.551357]  [<ffffffffa06bd28c>] __skb_fragment+0x25c/0x540 [tempesta_fw]
[52016.559145]  [<ffffffff810c4ef8>] ? __lock_acquire+0xec8/0x2030
[52016.565977]  [<ffffffffa06bd7d5>] ss_skb_cutoff_data+0x85/0x120 [tempesta_fw]
[52016.574036]  [<ffffffffa06a96ae>] __hdr_sub+0xfe/0x130 [tempesta_fw]
[52016.581329]  [<ffffffffa06aa0b7>] tfw_http_msg_hdr_xfrm+0x127/0x4b0 [tempesta_fw]
[52016.589749]  [<ffffffffa06aa11a>] ? tfw_http_msg_hdr_xfrm+0x18a/0x4b0 [tempesta_fw]
[52016.598352]  [<ffffffff810c4ef8>] ? __lock_acquire+0xec8/0x2030
[52016.605207]  [<ffffffffa06a77cb>] tfw_http_set_hdr_connection+0xab/0xc0 [tempesta_fw]
[52016.613962]  [<ffffffffa06a7829>] tfw_http_adjust_resp+0x49/0xb0 [tempesta_fw]
[52016.622119]  [<ffffffff810c05ff>] ? lock_release_holdtime.part.29+0xf/0x1f0
[52016.630001]  [<ffffffffa06a7df1>] tfw_http_resp_cache_cb+0x31/0xc0 [tempesta_fw]
[52016.638309]  [<ffffffffa06a3812>] tfw_cache_process+0xc2/0x140 [tempesta_fw]
[52016.646282]  [<ffffffffa06a760d>] ? tfw_http_popreq+0x3d/0x70 [tempesta_fw]
[52016.654176]  [<ffffffffa06a837e>] tfw_http_msg_process+0x23e/0x650 [tempesta_fw]
[52016.662517]  [<ffffffff81541cc5>] ? ip_queue_xmit+0x5/0x600
[52016.669048]  [<ffffffffa06a6e67>] __gfsm_fsm_exec+0x67/0xa0 [tempesta_fw]
[52016.676808]  [<ffffffffa06a70b4>] tfw_gfsm_dispatch+0x14/0x20 [tempesta_fw]
[52016.684745]  [<ffffffffa06a65ea>] tfw_connection_recv+0xaa/0xc0 [tempesta_fw]
[52016.692869]  [<ffffffffa06b8ef8>] ss_tcp_process_data+0x198/0x3d0 [tempesta_fw]
[52016.701160]  [<ffffffffa06b9670>] ss_tcp_data_ready+0x60/0xa0 [tempesta_fw]
[52016.709103]  [<ffffffff81557eda>] tcp_rcv_established+0x3fa/0x750
[52016.716164]  [<ffffffff81563613>] tcp_v4_do_rcv+0x103/0x440
[52016.722706]  [<ffffffff8156502d>] tcp_v4_rcv+0xa4d/0xcd0
[52016.728958]  [<ffffffff8153a299>] ? ip_local_deliver_finish+0x49/0x3f0
[52016.736410]  [<ffffffff8153a39a>] ip_local_deliver_finish+0x14a/0x3f0
[52016.743754]  [<ffffffff8153a299>] ? ip_local_deliver_finish+0x49/0x3f0
[52016.751158]  [<ffffffff8153a86a>] ip_local_deliver+0x4a/0xa0
[52016.757674]  [<ffffffff81539cf5>] ip_rcv_finish+0xb5/0x610
[52016.764059]  [<ffffffff8153ab57>] ip_rcv+0x297/0x3b0
[52016.769910]  [<ffffffff81539c40>] ? inet_del_offload+0x40/0x40
[52016.776605]  [<ffffffff814f2833>] __netif_receive_skb_core+0x233/0xca0
[52016.784004]  [<ffffffff814f3f75>] ? process_backlog+0x75/0x240
[52016.790682]  [<ffffffff814f32ba>] __netif_receive_skb+0x1a/0x80
[52016.797422]  [<ffffffff814f3f7e>] process_backlog+0x7e/0x240
[52016.803919]  [<ffffffff814f3fe3>] ? process_backlog+0xe3/0x240
[52016.810600]  [<ffffffff814f3c02>] net_rx_action+0x1f2/0x4f0
[52016.816991]  [<ffffffff8106d2a4>] __do_softirq+0xf4/0x630
[52016.823167]  [<ffffffff8106d7ff>] run_ksoftirqd+0x1f/0x60
[52016.829371]  [<ffffffff81091d57>] smpboot_thread_fn+0x187/0x280
[52016.836069]  [<ffffffff81091bd0>] ? sort_range+0x30/0x30
[52016.842160]  [<ffffffff8108db6f>] kthread+0x11f/0x140
[52016.847988]  [<ffffffff81621422>] ? wait_for_completion+0x102/0x130
[52016.854996]  [<ffffffff8108da50>] ? kthread_create_on_node+0x250/0x250
[52016.862199]  [<ffffffff81626a22>] ret_from_fork+0x42/0x70
[52016.868262]  [<ffffffff8108da50>] ? kthread_create_on_node+0x250/0x250
[52016.875455] Code: 00 00 41 0f b6 00 0f b6 c0 44 39 c8 0f 8f 28 02 00 00 4d 85 db 49 0f 45 db 8b 83 dc 00 00 00 48 8b 93 e0 00 00 00 48 8b 44 02 08 <8b> 88 94 00 00 00 39 88 90 00 00 00 0f 84 8d 00 00 00 48 8b 80
[52016.896759] RIP  [<ffffffffa06bccd9>] __split_pgfrag+0x5d9/0x930 [tempesta_fw]
[52016.904701]  RSP <ffff8800592b33e8>
[52016.908881] CR2: 0000000000000094
```

**Comment by keshonok** [comment]
This is actually a common problem in `ss_skb.c` that is currently present in the following functions:
```C
__extend_pgfrags()
__new_pgfrag()
__split_pgfrag_add()
__split_pgfrag_del()
```
In all of these functions `skb_shinfo(skb)->frag_list` is used unconditionally without checking whether it's valid or not. That may lead to dereferencing a NULL pointer.

**Comment by keshonok** [comment]
Fixed with 95e6b79.
